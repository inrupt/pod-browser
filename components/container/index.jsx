/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import React, { useMemo } from "react";
import T from "prop-types";
import { useTable, useSortBy } from "react-table";
import { createStyles, makeStyles } from "@material-ui/styles";
import { useBem } from "@solid/lit-prism-patterns";
import clsx from "clsx";

import ContainerTableRow, { renderResourceType } from "../containerTableRow";
import SortedTableCarat from "../sortedTableCarat";
import { useRedirectIfLoggedOut } from "../../src/effects/auth";
import { useFetchContainerResourceIris } from "../../src/hooks/solidClient";
import { getResourceName } from "../../src/solidClientHelpers/resource";

import Spinner from "../spinner";
import styles from "./styles";
import Breadcrumbs from "../breadcrumbs";
import PageHeader from "../containerPageHeader";
import ContainerDetails from "../containerDetails";
import { BookmarksContextProvider } from "../../src/contexts/bookmarksContext";
import { isHTTPError } from "../../src/solidClientHelpers/utils";
import AccessForbidden from "../accessForbidden";
import ResourceNotFound from "../resourceNotFound";

const useStyles = makeStyles((theme) => createStyles(styles(theme)));

export default function Container({ iri }) {
  useRedirectIfLoggedOut();
  const encodedIri = encodeURI(iri);

  const { data: resourceIris, mutate, error } = useFetchContainerResourceIris(
    encodedIri
  );

  const loading = typeof resourceIris === "undefined";

  const bem = useBem(useStyles());

  const columns = useMemo(
    () => [
      {
        header: "Icon",
        accessor: "icon",
        disableSortBy: true,
        modifiers: ["align-center", "width-preview"],
      },
      {
        header: "BookmarkIcon",
        accessor: "bookmark-icon",
        disableSortBy: true,
        modifiers: ["align-center", "width-preview"],
      },
      {
        Header: "File",
        accessor: "filename",
      },
      {
        Header: "Type",
        accessor: "type",
      },
    ],
    []
  );

  const data = useMemo(() => {
    if (!resourceIris) {
      return [];
    }

    return resourceIris.map((rIri) => ({
      iri: rIri,
      name: getResourceName(rIri),
      filename: getResourceName(rIri)?.toLowerCase(),
      type: renderResourceType(rIri),
    }));
  }, [resourceIris]);

  // TODO fix typescript errors below.
  const {
    getTableProps,
    getTableBodyProps,
    headerGroups,
    rows,
    prepareRow,
  } = useTable(
    {
      columns,
      data,
      defaultCanSort: true,
    },
    useSortBy
  );

  if (error && isHTTPError(error.message, 401)) return <AccessForbidden />;
  if (error && isHTTPError(error.message, 403)) return <AccessForbidden />;
  if (error && isHTTPError(error.message, 404)) return <ResourceNotFound />;
  if (error) throw error;

  if (loading) {
    return <Spinner />;
  }

  // react-table works through spreads.
  /* eslint react/jsx-props-no-spreading: 0 */
  return (
    <>
      <BookmarksContextProvider>
        <PageHeader mutate={mutate} resourceList={data} />
        <ContainerDetails mutate={mutate}>
          <Breadcrumbs />
          <table className={clsx(bem("table"))} {...getTableProps()}>
            <thead className={bem("table__header")}>
              {headerGroups.map((headerGroup) => (
                <tr
                  key={headerGroup.id}
                  className={bem("table__header-row")}
                  {...headerGroup.getHeaderGroupProps()}
                >
                  {headerGroup.headers.map((column) => (
                    <td
                      key={column.id}
                      className={bem(
                        "table__header-cell",
                        ...(column.modifiers || [])
                      )}
                      {...column.getHeaderProps(column.getSortByToggleProps())}
                    >
                      {column.render("Header")}
                      {` `}
                      <SortedTableCarat
                        sorted={column.isSorted}
                        sortedDesc={column.isSortedDesc}
                      />
                    </td>
                  ))}
                </tr>
              ))}
            </thead>
            <tbody className={bem("table__body")} {...getTableBodyProps()}>
              {!data || !data.length ? (
                <tr key="no-resources-found" className={bem("table__body-row")}>
                  <td rowSpan={3} className={bem("table__body-cell")}>
                    No resources were found within this container.
                  </td>
                </tr>
              ) : null}

              {rows.map((row) => {
                prepareRow(row);
                const details = row.original;
                return (
                  <ContainerTableRow key={details.iri} resource={details} />
                );
              })}
            </tbody>
          </table>
        </ContainerDetails>
      </BookmarksContextProvider>
    </>
  );
}

Container.propTypes = {
  iri: T.string.isRequired,
};
