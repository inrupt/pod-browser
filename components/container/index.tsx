/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

// @ts-nocheck
// react-table is super broken with sorting, so temporarily disable ts checking.

import { ReactElement, useEffect, useState, useContext, useMemo } from "react";
import { fetchLitDataset, getThingOne, getIriAll } from "lit-solid";
import { useTable, useSortBy, UseSortByOptions } from "react-table";
import {
  Table,
  TableBody,
  TableCell,
  TableHead,
  TableRow,
} from "@material-ui/core";

import DetailsContextMenu from "../detailsContextMenu";
import ContainerTableRow, { ResourceDetails } from "../containerTableRow";
import UserContext from "../../src/contexts/userContext";
import { useRedirectIfLoggedOut } from "../../src/effects/auth";
import {
  getIriPath,
  fetchResourceWithAcl,
  fetchFileWithAcl,
  namespace,
} from "../../src/lit-solid-helpers";
import Spinner from "../spinner";

export async function fetchResourceDetails(
  iri: string
): Promise<ResourceDetails> {
  const name = getIriPath(iri);
  let resource;
  try {
    resource = await fetchResourceWithAcl(iri);
  } catch (e) {
    resource = await fetchFileWithAcl(iri);
  }

  return {
    ...resource,
    name,
  };
}

export async function getResourceInfoFromContainerIri(
  containerIri: string
): Promise<ResourceDetails[]> {
  const litDataset = await fetchLitDataset(containerIri);
  const container = getThingOne(litDataset, containerIri);
  const iris = getIriAll(container, namespace.contains);

  const promises = iris.map(fetchResourceDetails).map((p) => p.catch((e) => e));
  return Promise.all(promises);
}

interface IPodList {
  iri: string;
}

export default function Container(props: IPodList): ReactElement {
  useRedirectIfLoggedOut();

  const { iri } = props;

  const { session, isLoadingSession } = useContext(UserContext);
  const [resources, setResources] = useState<ResourceDetails[]>([]);
  const [isLoading, setIsLoading] = useState(
    isLoadingSession || !resources || !resources.length
  );

  const columns = useMemo(
    () => [
      {
        Header: "File",
        accessor: "name",
      },
      {
        Header: "Type",
        accessor: "types[0]",
      },
      {
        Header: "My Access",
        accessor: "permissions[0].alias",
      },
    ],
    []
  );

  const data = useMemo(() => resources || [], [resources]);

  // TODO fix typescript errors below.
  const {
    getTableProps,
    getTableBodyProps,
    headerGroups,
    rows,
    prepareRow,
  } = useTable<UseSortByOptions<ResourceDetails>>(
    {
      columns,
      data,
    },
    useSortBy
  );

  useEffect(() => {
    if (!session) {
      return;
    }

    getResourceInfoFromContainerIri(iri)
      .then((loadedResources) => {
        setResources(loadedResources);
        setIsLoading(false);
      })
      .catch((e) => {
        throw e;
      });
  }, [session, iri]);

  if (isLoading) {
    return <Spinner />;
  }

  // react-table works through spreads.
  /* eslint react/jsx-props-no-spreading: 0 */
  return (
    <>
      <Table {...getTableProps()}>
        <TableHead>
          {headerGroups.map((headerGroup) => (
            <TableRow {...headerGroup.getHeaderGroupProps()}>
              {headerGroup.headers.map((column) => (
                <TableCell
                  {...column.getHeaderProps(column.getSortByToggleProps())}
                >
                  {column.render("Header")}
                  <span>
                    {column.isSorted && column.isSortedDesc ? "v" : ""}
                    {column.isSorted && !column.isSortedDesc ? "^" : ""}
                  </span>
                </TableCell>
              ))}
            </TableRow>
          ))}
        </TableHead>

        <TableBody {...getTableBodyProps()}>
          {!resources || !resources.length ? (
            <TableRow key="no-resources-found">
              <TableCell rowSpan={3}>
                No resources were found within this container.
              </TableCell>
            </TableRow>
          ) : null}

          {rows.map((row) => {
            prepareRow(row);
            const details = row.original as ResourceDetails;
            return <ContainerTableRow key={details.iri} resource={details} />;
          })}
        </TableBody>
      </Table>
      <DetailsContextMenu />
    </>
  );
}
