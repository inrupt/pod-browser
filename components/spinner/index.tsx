import React, { ReactElement } from "react";
import { CircularProgress, Container } from "@material-ui/core";
import { makeStyles } from "@material-ui/core/styles";

const useStyles = makeStyles(() => ({
  "MuiCicularProgress-root": {
    display: "block",
    position: "fixed",
    top: "50%",
    right: "50%",
  },
}));

export default function Spinner(): ReactElement {
  const classes = useStyles();
  return (
    <Container>
      <CircularProgress
        className={classes["MuiCicularProgress-root"]}
        color="primary"
      />
    </Container>
  );
}
