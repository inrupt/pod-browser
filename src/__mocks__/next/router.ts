export default {
  push: jest.fn().mockResolvedValue(null),
};
