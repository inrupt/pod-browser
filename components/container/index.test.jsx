/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import React from "react";
import * as RouterFns from "next/router";
import {
  addUrl,
  createThing,
  mockSolidDatasetFrom,
  setThing,
} from "@inrupt/solid-client";
import { space } from "rdf-namespaces";
import { renderWithTheme } from "../../__testUtils/withTheme";
import Container from "./index";
import useContainer from "../../src/hooks/useContainer";
import useAuthenticatedProfile from "../../src/hooks/useAuthenticatedProfile";
import { mockProfileAlice } from "../../__testUtils/mockPersonResource";
import usePodRootUri from "../../src/hooks/usePodRootUri";
import useResourceInfo from "../../src/hooks/useResourceInfo";
import useAccessControl from "../../src/hooks/useAccessControl";
import { getContainerResourceIris } from "../../src/models/container";
import { chain } from "../../src/solidClientHelpers/utils";
import { TESTCAFE_ID_AUTH_PROFILE_LOAD_ERROR } from "../authProfileLoadError";
import { TESTCAFE_ID_POD_ROOT_LOAD_ERROR } from "../podRootLoadError";
import { TESTCAFE_ID_NO_CONTROL_ERROR } from "../noControlWarning";
import { TESTCAFE_ID_ACCESS_FORBIDDEN } from "../accessForbidden";
import { TESTCAFE_ID_RESOURCE_NOT_FOUND } from "../resourceNotFound";
import { TESTCAFE_ID_NOT_SUPPORTED } from "../notSupported";
import { TESTCAFE_ID_SPINNER } from "../spinner";

jest.mock("../../src/hooks/useContainer");
const mockedContainerHook = useContainer;

jest.mock("../../src/models/container");
const mockedGetContainerResourceIris = getContainerResourceIris;

jest.mock("../../src/hooks/useAuthenticatedProfile");
const mockedAuthenticatedProfileHook = useAuthenticatedProfile;

jest.mock("../../src/hooks/usePodRootUri");
const mockedPodRootUriHook = usePodRootUri;

jest.mock("../../src/hooks/useResourceInfo");
const mockedResourceInfoHook = useResourceInfo;

jest.mock("../../src/hooks/useAccessControl");
const mockedAccessControlHook = useAccessControl;

describe("Container view", () => {
  const iri = "https://example.com/container/";
  const thing = createThing({ url: iri });
  const dataset = chain(mockSolidDatasetFrom(iri), (t) => setThing(t, thing));
  const container = { dataset, thing };

  beforeEach(() => {
    mockedContainerHook.mockReturnValue({ data: container, mutate: jest.fn() });
    mockedAuthenticatedProfileHook.mockReturnValue({
      data: mockProfileAlice((t) =>
        addUrl(t, space.storage, "https://example.com/")
      ),
    });
    mockedPodRootUriHook.mockReturnValue(iri);
    mockedResourceInfoHook.mockReturnValue({ data: container });
    mockedAccessControlHook.mockReturnValue({});
    jest.spyOn(RouterFns, "useRouter").mockReturnValue({
      asPath: "asPath",
      replace: jest.fn(),
      query: {},
    });
    mockedGetContainerResourceIris.mockReturnValue([
      "https://myaccount.mypodserver.com/inbox",
      "https://myaccount.mypodserver.com/private",
      "https://myaccount.mypodserver.com/note.txt",
    ]);
  });

  it("renders a table", () => {
    const { asFragment } = renderWithTheme(<Container iri={iri} />);
    expect(asFragment()).toMatchSnapshot();
  });

  test("Renders a spinner if data is loading", () => {
    mockedGetContainerResourceIris.mockReturnValue(undefined);

    const { asFragment, getByTestId } = renderWithTheme(
      <Container iri={iri} />
    );
    expect(asFragment()).toMatchSnapshot();
    expect(getByTestId(TESTCAFE_ID_SPINNER)).toBeDefined();
  });

  it("renders a spinner if iri is undefined (can happen during build step)", () => {
    const { getByTestId } = renderWithTheme(<Container iri={undefined} />);
    expect(getByTestId(TESTCAFE_ID_SPINNER)).toBeDefined();
  });

  it("renders Access Forbidden if the fetch for container returns 401", () => {
    mockedContainerHook.mockReturnValue({
      error: { message: "401" },
    });
    const { asFragment, getByTestId } = renderWithTheme(
      <Container iri={iri} />
    );
    expect(asFragment()).toMatchSnapshot();
    expect(getByTestId(TESTCAFE_ID_ACCESS_FORBIDDEN)).toBeDefined();
  });

  it("renders Access Forbidden if the fetch for container returns 403", () => {
    mockedContainerHook.mockReturnValue({
      error: { message: "403" },
    });
    const { asFragment, getByTestId } = renderWithTheme(
      <Container iri={iri} />
    );
    expect(asFragment()).toMatchSnapshot();
    expect(getByTestId(TESTCAFE_ID_ACCESS_FORBIDDEN)).toBeDefined();
  });

  it("renders Access Forbidden if the fetch for container returns 404", () => {
    mockedContainerHook.mockReturnValue({
      error: { message: "404" },
    });
    const { asFragment, getByTestId } = renderWithTheme(
      <Container iri={iri} />
    );
    expect(asFragment()).toMatchSnapshot();
    expect(getByTestId(TESTCAFE_ID_RESOURCE_NOT_FOUND)).toBeDefined();
  });

  it("renders Not supported if the fetch for container returns 500", () => {
    mockedContainerHook.mockReturnValue({
      error: { message: "500" },
    });
    const { asFragment, getByTestId } = renderWithTheme(
      <Container iri={iri} />
    );
    expect(asFragment()).toMatchSnapshot();
    expect(getByTestId(TESTCAFE_ID_NOT_SUPPORTED)).toBeDefined();
  });

  it("renders Not Supported if resource loaded is not a container", () => {
    const resourceIri = "http://example.com/resource";
    mockedContainerHook.mockReturnValue({
      data: { dataset: mockSolidDatasetFrom(resourceIri) },
      mutate: jest.fn(),
    });
    const { asFragment } = renderWithTheme(<Container iri={resourceIri} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders AuthProfileLoadError if fetching authenticated profile fails", () => {
    mockedAuthenticatedProfileHook.mockReturnValue({ error: new Error() });
    const { asFragment, getByTestId } = renderWithTheme(
      <Container iri={iri} />
    );
    expect(asFragment()).toMatchSnapshot();
    expect(getByTestId(TESTCAFE_ID_AUTH_PROFILE_LOAD_ERROR)).toBeDefined();
  });

  it("renders PodRootLoadError if it fails to fetch resourceInfo for podRoot", () => {
    mockedResourceInfoHook.mockReturnValue({ error: new Error() });
    const { asFragment, getByTestId } = renderWithTheme(
      <Container iri={iri} />
    );
    expect(asFragment()).toMatchSnapshot();
    expect(getByTestId(TESTCAFE_ID_POD_ROOT_LOAD_ERROR)).toBeDefined();
  });

  it("renders NoControlError if it fails to get access control for podRoot", () => {
    mockedAccessControlHook.mockReturnValue({ error: new Error() });
    const { asFragment, getByTestId } = renderWithTheme(
      <Container iri={iri} />
    );
    expect(asFragment()).toMatchSnapshot();
    expect(getByTestId(TESTCAFE_ID_NO_CONTROL_ERROR)).toBeDefined();
  });
});
