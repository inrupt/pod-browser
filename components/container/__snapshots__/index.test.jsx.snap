// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Container view Renders a spinner if data is loading 1`] = `
<DocumentFragment>
  <div
    class="PodBrowser-loading-indicator-container PodBrowser-loading-indicator-container--center"
  >
    <svg
      class="PodBrowser-loading-indicator"
      viewBox="22 22 44 44"
    >
      <circle
        cx="44"
        cy="44"
        fill="none"
        r="20.2"
        stroke-width="3.6"
      />
    </svg>
  </div>
</DocumentFragment>
`;

exports[`Container view Renders a table view without data 1`] = `
<DocumentFragment>
  <div
    class="PodBrowser-page-header PodBrowser-files-container-header"
  >
    <h1
      class="PodBrowser-page-header__title"
    >
      Files
    </h1>
    <div
      class="PodBrowser-page-header__content"
    />
    <div
      class="PodBrowser-page-header__toolbar"
    />
  </div>
  <div
    class="PodBrowser-drawer-container"
  >
    <div
      class="PodBrowser-drawer-container__content"
    >
      <div
        class="PodBrowser-sub-header"
      >
        <div
          class="PodBrowser-loading-indicator-container PodBrowser-loading-indicator-container--center"
        >
          <svg
            class="PodBrowser-loading-indicator"
            viewBox="22 22 44 44"
          >
            <circle
              cx="44"
              cy="44"
              fill="none"
              r="20.2"
              stroke-width="3.6"
            />
          </svg>
        </div>
        <div
          class="PodBrowser-actions-container"
        >
          <a
            class="PodBrowser-page-header__action"
            data-testid="folder-details-button"
            href="/resource/[iri]?action=details&resourceIri="
          >
            Folder Details
          </a>
          <button
            class="PodBrowser-page-header__action"
            data-testid="add-folder-button"
            type="button"
          >
            Create Folder
          </button>
          <label
            class="PodBrowser-page-header__action"
            data-testid="upload-file-button"
          >
            Upload File
            <input
              data-testid="upload-file-input"
              style="display: none;"
              type="file"
            />
          </label>
        </div>
      </div>
      <table
        class="PodBrowser-table"
        role="table"
      >
        <thead
          class="PodBrowser-table__header"
        >
          <tr
            class="PodBrowser-table__header-row"
            role="row"
          >
            <td
              class="PodBrowser-table__header-cell PodBrowser-table__header-cell--align-center PodBrowser-table__header-cell--width-preview"
              colspan="1"
              role="columnheader"
            >
                
            </td>
            <td
              class="PodBrowser-table__header-cell PodBrowser-table__header-cell--align-center PodBrowser-table__header-cell--width-preview"
              colspan="1"
              role="columnheader"
            >
                
            </td>
            <td
              class="PodBrowser-table__header-cell"
              colspan="1"
              role="columnheader"
              style="cursor: pointer;"
              title="Toggle SortBy"
            >
              File 
            </td>
            <td
              class="PodBrowser-table__header-cell"
              colspan="1"
              role="columnheader"
              style="cursor: pointer;"
              title="Toggle SortBy"
            >
              Type 
            </td>
          </tr>
        </thead>
        <tbody
          class="PodBrowser-table__body"
          role="rowgroup"
        >
          <tr
            class="PodBrowser-table__body-row"
          >
            <td
              class="PodBrowser-table__body-cell"
              rowspan="3"
            >
              No resources were found within this container.
            </td>
          </tr>
        </tbody>
      </table>
    </div>
    <section
      class="PodBrowser-drawer"
    >
      <div
        class="PodBrowser-drawer__content"
      >
        <div
          class="PodBrowser-drawer__content-wrapper"
        >
          <div
            class="PodBrowser-drawer__article"
          >
            <button
              class="PodBrowser-button PodBrowser-button--text"
            >
              <span
                class="PodBrowser-icon-cancel"
              />
               Close
            </button>
          </div>
          <section
            class="PodBrowser-centeredSection"
          >
            <h3
              class="PodBrowser-content-h3"
              title=""
            >
              null
            </h3>
          </section>
          <hr
            class="PodBrowser-root"
          />
          <div
            class="PodBrowser-loading-indicator-container PodBrowser-loading-indicator-container--center"
          >
            <svg
              class="PodBrowser-loading-indicator"
              viewBox="22 22 44 44"
            >
              <circle
                cx="44"
                cy="44"
                fill="none"
                r="20.2"
                stroke-width="3.6"
              />
            </svg>
          </div>
        </div>
      </div>
    </section>
  </div>
</DocumentFragment>
`;

exports[`Container view Renders a table with data 1`] = `
<DocumentFragment>
  <div
    class="PodBrowser-page-header PodBrowser-files-container-header"
  >
    <h1
      class="PodBrowser-page-header__title"
    >
      Files
    </h1>
    <div
      class="PodBrowser-page-header__content"
    />
    <div
      class="PodBrowser-page-header__toolbar"
    />
  </div>
  <div
    class="PodBrowser-drawer-container"
  >
    <div
      class="PodBrowser-drawer-container__content"
    >
      <div
        class="PodBrowser-sub-header"
      >
        <div
          class="PodBrowser-loading-indicator-container PodBrowser-loading-indicator-container--center"
        >
          <svg
            class="PodBrowser-loading-indicator"
            viewBox="22 22 44 44"
          >
            <circle
              cx="44"
              cy="44"
              fill="none"
              r="20.2"
              stroke-width="3.6"
            />
          </svg>
        </div>
        <div
          class="PodBrowser-actions-container"
        >
          <a
            class="PodBrowser-page-header__action"
            data-testid="folder-details-button"
            href="/resource/[iri]?action=details&resourceIri="
          >
            Folder Details
          </a>
          <button
            class="PodBrowser-page-header__action"
            data-testid="add-folder-button"
            type="button"
          >
            Create Folder
          </button>
          <label
            class="PodBrowser-page-header__action"
            data-testid="upload-file-button"
          >
            Upload File
            <input
              data-testid="upload-file-input"
              style="display: none;"
              type="file"
            />
          </label>
        </div>
      </div>
      <table
        class="PodBrowser-table"
        role="table"
      >
        <thead
          class="PodBrowser-table__header"
        >
          <tr
            class="PodBrowser-table__header-row"
            role="row"
          >
            <td
              class="PodBrowser-table__header-cell PodBrowser-table__header-cell--align-center PodBrowser-table__header-cell--width-preview"
              colspan="1"
              role="columnheader"
            >
                
            </td>
            <td
              class="PodBrowser-table__header-cell PodBrowser-table__header-cell--align-center PodBrowser-table__header-cell--width-preview"
              colspan="1"
              role="columnheader"
            >
                
            </td>
            <td
              class="PodBrowser-table__header-cell"
              colspan="1"
              role="columnheader"
              style="cursor: pointer;"
              title="Toggle SortBy"
            >
              File 
            </td>
            <td
              class="PodBrowser-table__header-cell"
              colspan="1"
              role="columnheader"
              style="cursor: pointer;"
              title="Toggle SortBy"
            >
              Type 
            </td>
          </tr>
        </thead>
        <tbody
          class="PodBrowser-table__body"
          role="rowgroup"
        >
          <tr
            class="PodBrowser-table__body-row PodBrowser-table__body-row--selectable"
            tabindex="0"
          >
            <td
              class="PodBrowser-table__body-cell PodBrowser-table__body-cell--align-center PodBrowser-table__body-cell--width-preview"
            >
              <button
                class="PodBrowser-button PodBrowser-button--text PodBrowser-bookmark-button"
                data-testid="bookmark-button"
                disabled=""
                type="button"
              >
                <i
                  alt="bookmark button"
                  aria-label="bookmark"
                  class="PodBrowser-icon-star PodBrowser-bookmark-icon-unselected"
                />
              </button>
            </td>
            <td
              class="PodBrowser-table__body-cell PodBrowser-table__body-cell--align-center PodBrowser-table__body-cell--width-preview"
            >
              <i
                class="PodBrowser-icon-file PodBrowser-resource-icon"
              />
            </td>
            <td
              class="PodBrowser-table__body-cell"
            >
              inbox
            </td>
            <td
              class="PodBrowser-table__body-cell"
            >
              Resource
            </td>
          </tr>
          <tr
            class="PodBrowser-table__body-row PodBrowser-table__body-row--selectable"
            tabindex="0"
          >
            <td
              class="PodBrowser-table__body-cell PodBrowser-table__body-cell--align-center PodBrowser-table__body-cell--width-preview"
            >
              <button
                class="PodBrowser-button PodBrowser-button--text PodBrowser-bookmark-button"
                data-testid="bookmark-button"
                disabled=""
                type="button"
              >
                <i
                  alt="bookmark button"
                  aria-label="bookmark"
                  class="PodBrowser-icon-star PodBrowser-bookmark-icon-unselected"
                />
              </button>
            </td>
            <td
              class="PodBrowser-table__body-cell PodBrowser-table__body-cell--align-center PodBrowser-table__body-cell--width-preview"
            >
              <i
                class="PodBrowser-icon-file PodBrowser-resource-icon"
              />
            </td>
            <td
              class="PodBrowser-table__body-cell"
            >
              private
            </td>
            <td
              class="PodBrowser-table__body-cell"
            >
              Resource
            </td>
          </tr>
          <tr
            class="PodBrowser-table__body-row PodBrowser-table__body-row--selectable"
            tabindex="0"
          >
            <td
              class="PodBrowser-table__body-cell PodBrowser-table__body-cell--align-center PodBrowser-table__body-cell--width-preview"
            >
              <button
                class="PodBrowser-button PodBrowser-button--text PodBrowser-bookmark-button"
                data-testid="bookmark-button"
                disabled=""
                type="button"
              >
                <i
                  alt="bookmark button"
                  aria-label="bookmark"
                  class="PodBrowser-icon-star PodBrowser-bookmark-icon-unselected"
                />
              </button>
            </td>
            <td
              class="PodBrowser-table__body-cell PodBrowser-table__body-cell--align-center PodBrowser-table__body-cell--width-preview"
            >
              <i
                class="PodBrowser-icon-file PodBrowser-resource-icon"
              />
            </td>
            <td
              class="PodBrowser-table__body-cell"
            >
              note.txt
            </td>
            <td
              class="PodBrowser-table__body-cell"
            >
              Resource
            </td>
          </tr>
        </tbody>
      </table>
    </div>
    <section
      class="PodBrowser-drawer"
    >
      <div
        class="PodBrowser-drawer__content"
      >
        <div
          class="PodBrowser-drawer__content-wrapper"
        >
          <div
            class="PodBrowser-drawer__article"
          >
            <button
              class="PodBrowser-button PodBrowser-button--text"
            >
              <span
                class="PodBrowser-icon-cancel"
              />
               Close
            </button>
          </div>
          <section
            class="PodBrowser-centeredSection"
          >
            <h3
              class="PodBrowser-content-h3"
              title=""
            >
              null
            </h3>
          </section>
          <hr
            class="PodBrowser-root"
          />
          <div
            class="PodBrowser-loading-indicator-container PodBrowser-loading-indicator-container--center"
          >
            <svg
              class="PodBrowser-loading-indicator"
              viewBox="22 22 44 44"
            >
              <circle
                cx="44"
                cy="44"
                fill="none"
                r="20.2"
                stroke-width="3.6"
              />
            </svg>
          </div>
        </div>
      </div>
    </section>
  </div>
</DocumentFragment>
`;

exports[`Container view renders Access Forbidden if the fetch for container returns 401 1`] = `
<DocumentFragment>
  <div
    class="PodBrowser-loading-indicator-container PodBrowser-loading-indicator-container--center"
  >
    <svg
      class="PodBrowser-loading-indicator"
      viewBox="22 22 44 44"
    >
      <circle
        cx="44"
        cy="44"
        fill="none"
        r="20.2"
        stroke-width="3.6"
      />
    </svg>
  </div>
</DocumentFragment>
`;

exports[`Container view renders Access Forbidden if the fetch for container returns 403 1`] = `
<DocumentFragment>
  <div
    class="PodBrowser-loading-indicator-container PodBrowser-loading-indicator-container--center"
  >
    <svg
      class="PodBrowser-loading-indicator"
      viewBox="22 22 44 44"
    >
      <circle
        cx="44"
        cy="44"
        fill="none"
        r="20.2"
        stroke-width="3.6"
      />
    </svg>
  </div>
</DocumentFragment>
`;

exports[`Container view renders Access Forbidden if the fetch for container returns 404 1`] = `
<DocumentFragment>
  <div
    class="PodBrowser-loading-indicator-container PodBrowser-loading-indicator-container--center"
  >
    <svg
      class="PodBrowser-loading-indicator"
      viewBox="22 22 44 44"
    >
      <circle
        cx="44"
        cy="44"
        fill="none"
        r="20.2"
        stroke-width="3.6"
      />
    </svg>
  </div>
</DocumentFragment>
`;

exports[`Container view renders Not Supported if resource loaded is not a container 1`] = `
<DocumentFragment>
  <div
    class="PodBrowser-page-header"
  >
    <h1
      class="PodBrowser-page-header__title"
    >
      Resource Not Supported
    </h1>
    <div
      class="PodBrowser-page-header__content"
    />
    <div
      class="PodBrowser-page-header__toolbar"
    />
  </div>
  <div
    class="PodBrowser-container"
  >
    <div
      class="PodBrowser-content"
    >
      <p>
        PodBrowser can only show resources that are Solid containers.
      </p>
      <p>
        <span>
          Navigate to the resource directly (this will navigate away from PodBrowser): 
        </span>
        <a />
      </p>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Container view renders Not supported if the fetch for container returns 500 1`] = `
<DocumentFragment>
  <div
    class="PodBrowser-loading-indicator-container PodBrowser-loading-indicator-container--center"
  >
    <svg
      class="PodBrowser-loading-indicator"
      viewBox="22 22 44 44"
    >
      <circle
        cx="44"
        cy="44"
        fill="none"
        r="20.2"
        stroke-width="3.6"
      />
    </svg>
  </div>
</DocumentFragment>
`;
